### ISSUE TITLE
Fix integration with the new "My BMW" app

### ISSUE THREAD

**Issue by dimakaob** [hufman/AndroidAutoIdrive#239]
Hello! There is this new application "My bmw" instead of the old "Connected". 
Are you planning to update AndroidAutoidrive?  
Thank you.

**Comment by hufman** [comment]
Good morning!
This app requires two main pieces of functionality from the Connected app: The Apps tunnel to connect to the car, and the Security module to authenticate to the car.
"My BMW" has its security module in a slightly different place, so I added support for it on [November 23rd](https://github.com/hufman/AndroidAutoIdrive/commit/62688a17bc1ce0c76e5dbb6fb6dee636599c9a96)
The app will use whichever Apps tunnel starts up on the phone. At that time, My BMW didn't enable the Apps connection to the car, but I'm hearing reports that it works now. If My BMW does start the Apps tunnel, this app should automatically use it.
Any of the automatic builds since November should include support, if the underlying Apps tunnel is present. If you are running the last Release from August, try out the [latest build](https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-master-nomap-sentry-release.apk) instead.

**Comment by fabssor** [comment]
So for me, it is working fine with the MyBMW app. 
But for now, as described in #238 you need to start AndroidAutoIDrive manually each time. Spotify etc. should then show up.

**Comment by LACNR** [comment]
Since the latest Update for MyBMW the two older Connected Apps don't seem to acknowledge the connection to the car, my phone doesn't either and even on manual start up no connection can be made, not even AAiD App. In car when I tried to manually run the connection assistant it tells me I have to download the Connected App, which I of course have.

I deinstalled MyBMW and restarted my phone but nothing seems to help, I believe the latest Update of MyBMW disabled the old apps, because when starting the previous Connected App it tells that the App will be deactivated on June 30th.

Anyone having similar trouble?

**Comment by fabssor** [comment]
Be sure that you have only installed one BMW App. For you this means onyl install the MyBMW App and deinstall the "old" Connected App. Greetings!

**Comment by LACNR** [comment]
@fabssor 

It worked always fine having all three BMW installed before until I updated MyBMW yesterday, because the MyBMW offers not much functionality it is basically useless for me or does it also have Calendar and more like the previous Connected Apps including the classic?

**Comment by fabssor** [comment]
Ah no I also noticed today that the calendar feature is nor supported by MyBMW.

**Comment by LACNR** [comment]
I was eventually able to fix it, I deinstalled the previous Connected App and then reinstalled it upon first connection I had to select the default App either Connected or Connected Classic.

I selected as usually the later Connected App, because it used to connect with both Apps anyways, but it didn't work, then I choose to delete the default startup and reconnected, this time I selected the Connected Classic App as default and it immediatly connected to the car, everything functional from the Connected Classic (News, Audioplayer, Calendar, Wiki Local, ECO Pro Analyser) app and the apps supported by AAiDrive.
The latest Connected App is not connecting though yet.

**Comment by jof284** [comment]
Seems there is like a built-in verification process in the latest BMW and MINI Connected apps where it seems that that verification fails (maybe if the date is later than a defined date) it refuses to connect. The way you can workaround this for now is with the latest Connected app stop the app, clear the cache and data and then put the phone in airplane mode, connect to the car  (I tested via USB) and it should connect just fine. The first time you open the Connected app it will refuse to connect again so I would just leave it without credentials.

**Comment by dimakaob** [comment]
Hi there. 
Today at the morning tried this latest build. Works fine so far. 
Will try later at the evening. 
Thank you! 

P.S. So glad you fixed the album icon. Different background color from August release freaked me out. %D

**Comment by dimakaob** [comment]
Well, just tested again. Everything works just perfect!
And I noticed, that now my car can see the messages from Instagram, or from Google. Wow!!!
 Is there any chance to get Whatsapp or Viber messages?

**Comment by LACNR** [comment]
I can use AutoiDrive as usual before but only with the Connected Classic App which isn't available on the app stores anymore, the newer Connected App doesn't seem to recognize the car anymore and vice versa nor via Bluetooth/USB, the MyBMW does notify that Bluetooth Connection to the car has been made but due to the lack functionality I don't see if the MyBMW app has an affect and functions over the original Connected Classic app.

My assume is, the current Connected App is now no longer functional in the car? I will ask my friends if they have similar problems but also ask in the BMW forums.

@dimakaob I could read and see all kinds of messages such WhatsApp, Insta, Facebook, Weather, Messenger and so on since at least the August release (first time I got this AAiDrive super awesome app). So it should work for you too, although I installed the new build yesterday things might have changed.

**Comment by hufman** [comment]
@LACNR I'm guessing that BMW changed a server-side setting and flipped everyone's Connected 6.5 to this promotional mode, disabling the car app functionality this week. There's a link in #238 to download the previous Connected 6.4, which should restore the functionality for you.
The BMW app's notification displays if a Bluetooth Music connection is made to a device with a name that starts with BMW or MINI (perhaps also verifying that the Bluetooth provides the SPP support for the car protocol), but is otherwise unrelated to the status of the car app connection. In my tests, this notification remains if you uncheck the Apps option in the car. For further information, you can look at the advanced Connection tab in the app, which shows performance details about the BCL Tunnel, including whether it is actively running.
From what I'm seeing in the code, MyBMW now only includes a car application to read data from the CDS, perhaps providing the car information data for cars that aren't cloud connected. And so, the only real clue that MyBMW is connected over the app protocol is that this BCL Tunnel Statistics panel shows active data.

@dimakaob AAI has supported displaying your phone notifications (including messages) in the car for a long time (via the unlabeled book icon). However, this current version has a ton of improvements in this area, such as adding new messages to an ID5 car's Notification Center, which is a little more prominent and displays the statusbar icon if the app decided to not trigger a popup for the notification. I'm glad you like it :)

**Comment by dimakaob** [comment]
But still i can't get notifications from Whatsapp and Viber =((( Can you tell me where i should check? 
I have 
1. Access to notifications
2. Access to SMS notifications
   - pop up notifications
        - pop ip notifications with the passenger (what is it?)
3. Play the notification sound
4. Speak the displayed notification
5. Speak new notifications
and fast replies

**Comment by LACNR** [comment]
@hufman 

Thank you :) , too bad that BMW did that, but I will try out the 6.4 and downloaded it for later.

@dimakaob It should work right away when you have enabled all notifications at least 1. and 2.
I think Pop up notifications with passenger is when your BMW has the option enhanced bluetooth connectivity allowing two phones being paired to the car via bluetooth at the same time to receive calls, play music and so on. You will also a second microphone on the passenger side.

**Comment by dimakaob** [comment]
@LACNR 
Thank you!
It's funny, but here is what i did. 
I started my car and ask my friend to send me an instagram message. Notification worked fine. Then i asked him to send me a Viber message. Nothing. The radio was on. Then i switched to Spotify. And then asked again. And i got the Viber message! Then i switched back to radio and it still works fine. Hooray!

**Comment by hufman** [comment]
@dimakaob Where are you looking for these messages? If you go into the unlabeled book icon in the Online Services menu, you should see a live-updating list of all notifications currently displayed on your phone. If the Viber message was received, and is showing on the phone's notification list, it should show up here right away.

@LACNR The Show Popups With Passenger option is something simpler than that: The car tells the app whether someone is sitting in the front passenger seat, and this option is used to show or not-show new notification popups while a passenger is detected.

**Comment by DominikS66** [comment]
@hufman on my car (F20 2019, ID6) AndroidAutoIdrive stopped working this week due to the issue you mentiond before. Solution which works for me at this time: I installed both MyBMW and BMW Connected App on Smartphone (Samsung S9), but never started BMW Connected App, only MyBMW to connect to the car. With this workaround, the actual AndroidAutoIDrive release (1.2.2) is working again. But it would be nice to get a new release of AndroidAutoIDrive soon (or I should make my own apk..)

**Comment by LACNR** [comment]
@DominikS66

Hey 😉

If you download the latest build https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-master-nomap-sentry-release.apk MyBMW should work on its own but it doesn't launch AutoiDrive, you will need to open it manually, whenever you establish a connection.

Which brings me to the question, with you not being logged in in the Connected App, does AutoiDrive launches on its own when using MyBMW or do you have to open it yourself.

**Comment by dimakaob** [comment]
@LACNR I don't launch anything manually. Just start the engine.

**Comment by LACNR** [comment]
@dimakaob When you use the Connected Apps they start and launch AutoiDrive automatically, but people reported (and as hufman is stating in his description) that MyBMW doesn't launch AAiD automatically, so I assumed that is still the case. Apparently, using MyBMW as the main, and Connected only being installed but not logged in it works automatically, if anyone can approve on that?

**Comment by DominikS66** [comment]
@LACNR AndroidAutoIDrive launches automatic at my Smartphone with this workaround. But after installing the latest master it doesn't automatic start AutoIDrive any more, even after uninstalling and installing AndroidAutoIDrive 1.2.2 Release and BMW Connected App...

**Comment by dimakaob** [comment]
I have only the new My bmw app and the latest build AAI. Works great without any movements..

**Comment by LACNR** [comment]
Going back to Version 6.4- of Connected does re-establish the connectivity with the Connected App and it works in cooperation with Connected Classic. MyBMW might will be attractive once it becomes more compatible with older model years and adds at least internal apps such as the BMW Calendar. MyBMW does recognize a connection to my BMW just fine but I can't tell if it is functional over the Connected apps.

But we are heading somewhere and there some solutions at this point.

**Comment by TBGR** [comment]
MyBMW does not work for me. Probably because it asks for a VIN to add the vehicle and my car came out of the factory with a CCC headunit.

**Comment by DominikS66** [comment]
After installing latest built (from tonight ?) and grant access to notifications for AAID, automatic start of AAID works with the MyBMW app.

**Comment by TBGR** [comment]
> After installing latest built (from tonight ?) and grant access to notifications for AAID, automatic start of AAID works with the MyBMW app.

Probably, but the app connection does not start. Always running the latest build.

**Comment by daniel1v** [comment]
> I have only the new My bmw app and the latest build AAI. Works great without any movements..

For me it is the same. I am using My BMW  with latest AndroidAutoIdrive since a couple of days and it works perfectly. Everything starts and connects with the phone inside my pocket. Even after a phone restart and without any manual starting of My BMW or AndroidAutoIdrive. The  connection is established a few seconds after the bluetooth connection and it is very stable.

Maybe everyone should double check to disable battery optimization for the new My BMW App.

**Comment by hufman** [comment]
I haven't fixed anything yet 😅 Still working on it!

**Comment by daniel1v** [comment]
@hufman yes, I know, I have always an eye on your latest commits. 😀

Just wanted to let you know, that there are indeed cases, where it works and that there is maybe (?) nothing to fix on your side.

At least for my HW/SW combination, the whole connection process works automatically and reliably, without unlocking the phone.

**Comment by LACNR** [comment]
Could it be that BMW's after MY2017 don't require an USB connection for the BMW Apps connectivity therefor MyBMW might be only optimized for Bluetooth connection right now, models before MY2017 require an USB connection to have BMW Apps, while MyBMW does detect the Bluetooth connection to your BMW regardless the model year it might not be able to make AAID launch itself.

I haven't tried it with MyBMW only, but maybe others can fill in the gap.

**Comment by DominikS66** [comment]
@hufman this is an interesting issue that you didn't change anything...for your information, there's a binary difference in apk file content between the file I've downloaded on Saturday morning 10:43 CEST and Sunday morning 07:22 CEST, using the link https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-master-nomap-sentry-release.apk:
![grafik](https://user-images.githubusercontent.com/81849700/113540173-dba1ed00-95df-11eb-9d26-12512df038cc.png)

**Comment by hufman** [comment]
Yes, I merged in an updated Spanish translation #242, which wouldn't have any impact on the connection logic :)

**Comment by DominikS66** [comment]
@hufman ok now I tested the security setting for notifications access (the second thing I've changed yesterday). And I found out that on my Galaxy S9 automatic start of AAID on bluetooth connection to car only works if AAID has been granted access rights to notifications. This behavior is reproducible on my Smartphone. Maybe this is an important information for other users here.

**Comment by daniel1v** [comment]
> I haven't tried it with MyBMW only, but maybe others can fill in the gap.

So when you write, you did not try it with My BMW *only*, do you mean, you had BMW Connected *and* MyBMW installed? 

I am asking this, because I remember there were issues in the past, where both apps 'fought' for the connection what lead to connection issues.

**Comment by LACNR** [comment]
> > I haven't tried it with MyBMW only, but maybe others can fill in the gap.
> 
> So when you write, you did not try it with My BMW _only_, do you mean, you had BMW Connected _and_ MyBMW installed?
> 
> 

Yes, I have all three apps, Connected, Connected Classic and MyBMW installed because primarily I'm focused on using Connected Classic (and BMW Connected) as of now because MyBMW doesn't show me car data in the app, also it doesn't have the extra BMW apps such as the calendar.
That's why I didn't test out MyBMW alone, although when I connect it to my car all three BMW apps give the notification that they are connected to the BMW.

**Comment by hufman** [comment]
"Force Audio Focus via USB" means to tell the car to listen to the USB apps connection for audio, even if the phone doesn't seem to support outputting audio through it. The car will behave like the app has audio focus, and the Media hardware button will go back to the Audioplayer app, but the music will likely come out of your phone speakers.

**Comment by hufman** [comment]
I've uploaded a [build to test a workaround](https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-bt_launch-nomap-sentry-release.apk). This method attempts to start the app's MainService when the Bluetooth connection to the car connects, and then it will do the same probing process as when opening the UI to manually start the connection.
I'd greatly appreciate your assistance in testing this approach, especially if you are using only MyBMW and the automatic connection isn't working for you, especially if you didn't enable Notification Access (which includes a separate workaround already). The method I'm using to start the MainService from the background needs me to show a notification, and I technically do, but then hide it right away until the connection is finalized, and so I'd like confirmation that this approach works well.

So far, it feels like My BMW connects the Bluetooth Apps to the car faster than BMW Connected, has anyone else noticed this too?

**Comment by daniel1v** [comment]
> I've uploaded a [build to test a workaround](https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-bt_launch-nomap-sentry-release.apk). This method attempts to start the app's MainService when the Bluetooth connection to the car connects, and then it will do the same probing process as when opening the UI to manually start the connection.
> I'd greatly appreciate your assistance in testing this approach, especially if you are using only MyBMW and the automatic connection isn't working for you, especially if you didn't enable Notification Access (which includes a separate workaround already). The method I'm using to start the MainService from the background needs me to show a notification, and I technically do, but then hide it right away until the connection is finalized, and so I'd like confirmation that this approach works well.
> 
> So far, it feels like My BMW connects the Bluetooth Apps to the car faster than BMW Connected, has anyone else noticed this too?

I torally can confirm this. It connects faster and is also more reliably and stable, then the Connected App.

**Comment by fabssor** [comment]
> I've uploaded a [build to test a workaround](https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-bt_launch-nomap-sentry-release.apk). This method attempts to start the app's MainService when the Bluetooth connection to the car connects, and then it will do the same probing process as when opening the UI to manually start the connection.
> I'd greatly appreciate your assistance in testing this approach, especially if you are using only MyBMW and the automatic connection isn't working for you, especially if you didn't enable Notification Access (which includes a separate workaround already). The method I'm using to start the MainService from the background needs me to show a notification, and I technically do, but then hide it right away until the connection is finalized, and so I'd like confirmation that this approach works well.
> 
> So far, it feels like My BMW connects the Bluetooth Apps to the car faster than BMW Connected, has anyone else

Works like a charm! Thank you so much Walter! 
Yes I also noticed that the Bluetooth Apps connect faster. Maybe because the BMW Apps (like calender) are not included?!

**Comment by DominikS66** [comment]
> 
> I've uploaded a [build to test a workaround](https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-bt_launch-nomap-sentry-release.apk). This method[...]
> So far, it feels like My BMW connects the Bluetooth Apps to the car faster than BMW Connected, has anyone else noticed this too?

I've tested the built and AAID is started automatic now no matter if access to nofications is granted or not. So your changes are working.
And I can also confirm that the MyBMW app is connecting faster via bluetooth and it seems the bluetooth connection is working more reliable than with BMW connected app.

**Comment by LACNR** [comment]
Do you all require an USB connection for the BMW Apps integration (models before 2017) or do you have the bluetooth apps integration and only need to connect via bluetooth to have Spotify, Apple Music,... when using Connected and specifically for this case MyBMW?

The question is if MyBMW supports apps integration via USB which was needed for models before 2017 where you have to connect your phone via USB to be able to use the BMW Apps and other third-party-apps.

**Comment by daniel1v** [comment]
> Do you all require an USB connection for the BMW Apps integration (models before 2017) or do you have the bluetooth apps integration and only need to connect via bluetooth to have Spotify, Apple Music,... when using Connected and specifically for this case MyBMW?
> 
> The question is if MyBMW supports apps integration via USB which was needed for models before 2017 where you have to connect your phone via USB to be able to use the BMW Apps and other third-party-apps.

I have a newer model with blueooth app support.

**Comment by masteralarm** [comment]
> The question is if MyBMW supports apps integration via USB which was needed for models before 2017 where you have to connect your phone via USB to be able to use the BMW Apps and other third-party-apps.

I have an ID4 car where the apps connections is only supported over USB and I tested it over the weekend. The new MyBMW app seems to not support the USB connection at all (easy to identify: when I plug the phone in, it is not suggested by the Android USB connection dialog). The old Connected app did also drop/deactivate support and shows the "switch to the new app" information.

I needed to switch back to Connected 6.4. Now everything works again.

**Comment by LACNR** [comment]
> > The question is if MyBMW supports apps integration via USB which was needed for models before 2017 where you have to connect your phone via USB to be able to use the BMW Apps and other third-party-apps.
> 
> I have an ID4 car where the apps connections is only supported over USB and I tested it over the weekend. The new MyBMW app seems to not support the USB connection at all (easy to identify: when I plug the phone in, it is not suggested by the Android USB connection dialog). The old Connected app did also drop/deactivate support and shows the "switch to the new app" information.
> 
> I needed to switch back to Connected 6.4. Now everything works again.

I thought so, that was my fear. Even, going back to the 6.4 Version is only a workaround until June 30th, when the Connected App will deactivated completely, I assume the only longterm fix would be switching to the Connected Classic app which is not using a login to BMW servers. Maybe, we can get BMW to add USB connectivity.

Although MyBMW does recognize the car when connected via Bluetooth at least it says Connected with BMW in the notifications tab, I fear it doesn't matter much to the car and the functions. Did the USB + Bluetooth combo with the latest build help out?

**Comment by desertspotter** [comment]
I can also confirm that the new build is working fine (MY21 F22, ID6) without notification access, and no bmw connected app installed. My only issue now is #217 , AAI doen't close Spotify connection, but just in case, the app works fine 

Regarding USB and my app, I cannot test that case in my car, but for me, connected 6.4 was functional over the weekend without login in the app, at least my user/pass was not recognized but the app enabled app functionality and establish connection with AAI. I guess that if BMW don't release app over usb functionality before June you'll be still able to use 6.4 without login.

**Comment by hufman** [comment]
Thank you so much for the feedback! I'll keep this in mind as I phrase the new text and buttons to complete MyBMW support.

I believe, though would like confirmation before officially recommending it, that earlier versions of BMW Connected are able to launch the car connection even before logging in to Connected Drive. 
 (As I type this, I see desertspotter's comment suggesting that 6.4 works in this way) Even if they disable logging in from BMW Connected to shut down the older app, this may be a workable solution for ID4 users. Does ID4 even still support the remote functionality that MyBMW is based around.
Connected Classic seemed to interact poorly with modern Android's process killer, though it otherwise has some great functionality, and I definitely intend to continue supporting it!

**Comment by LACNR** [comment]
Yeah, if that would still work even if they shut down the Connected App that would be great although it is a chance to notify BMW that they need to add USB compatibility to read car data such as fuel, mileage and total distance - there are hundred thousands of cars that rely on the USB connectivity and they even say the App is optimized for vehicles since 2014.

True, I have to disable Connected Classic via apps manager because it needs alot of battery after usage but otherwise as you said great functionality.

Yes, ID4 (MY2016) still supports the remote services just reordered the service today. Although, BMW Online might not be functional towards the end of this year (germany) as providers planned to shutdown 3G, unless it runs on 4G despite saying 3G in the upper left corner.

**Comment by hufman** [comment]
I updated [this branch's build](https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-bt_launch-nomap-sentry-release.apk) to improve the UI text about MyBMW being supported, check it out! I think it's pretty much ready to merge, if it all checks out.

**Comment by LACNR** [comment]
I did some testing and I can confirm as well that if you go back to Connected 6.4 and do not login into the app with your account, the car will still recognize the app and connect with the car to use the Calendar, BMW Connected Destinations and of course the added apps via AAiD.

I also noticed MyBMW does recognize and notifies you on the phone with a PopUp that it is connected to your BMW when you connect via USB only, meaning it does support USB and Bluetooth connection independend from each other, although it doesn't seem to be able reading the car data via USB nor Bluetooth, unless it reads it real-time via on-the-air for newer models.

**Comment by hufman** [comment]
Interesting! When I tested with the My Mini app, I did notice that the phone flipped the USB connection to Accessory mode, so the Mini app is successfully opening the accessory connection, but seemingly not launching the BCL tunnel over it. The car also doesn't indicate that the Apps connection is running.
The popup you see might be because the popup seems to show for any Bluetooth connection to the car. Even if I disconnect Music and Apps, the popup still shows up. Just now in my test, I kept the phone's Bluetooth entirely disabled and didn't see the popup.

In this recent branch update, I added text to indicate that USB mode is not supported by MyBMW. I think I should go forward with this change, which should be very easy to change if the USB support changes. I do see that apkpure has a new minor MyBMW update from yesterday, so I'll wait a bit to hear confirmation on if USB support changed.

Thank you for your help in testing!

**Comment by LACNR** [comment]
Yeah pretty interesting, I disabled bluetooth and the BMW Connected popup disappeared, but MyBMW popup remained, then I also disconnected the USB connection and the MyBMW popup disappeared and when I reconnected it via USB the MyBMW PopUp appeared instantly again. I repeated this to make sure it isn't a malfunction from the previous session, but apparently MyBMW does support USB connection to the car even if it doesn't have much use right now or doesn't work properly I can't tell.

I have long forgotten that later versions of BMW Connected only show the popup when connected via bluetooth saying I can use the calendar, search my destinations and other apps - of course in my case to use them I need to connect via USB regardless the popup. But MyBMW shows the pop up with either connection.

Yes, I updated MyBMW off the play store yesterday, but I haven't noticed any improvements but to be honest I didn't test it further.

Thank you for all this and I will try out the latest branch tomorrow :)

**Comment by DominikS66** [comment]
@hufman you'd closed this issue...but I have some reports from testing yesterday. The bt_launch branch works on my phone arround 20 minutes after starting the car, then for some reason the apps connection is broken. Tested this with playing Spotify music while driving the car. This happens 4 times now. To bring back connection I need to kill AAId and Spotify on my phone. This issue doesn't occur with the branch without bt_launch (Version 1.2.2.404-bdec840), I downgraded to this version and the problem dissappeared. I've also tried a homemade apk taken from the branch bt_launch, same problem here, after arround 20 minutes AAId stops working.

**Comment by hufman** [comment]
Interesting! I did see this twice, but both times happening about 30 seconds from getting home so I didn't check further. I did notice that AndroidAutoIdrive didn't start up it's notification saying it's connected, and I wonder if that's related. Do you observe whether AAI's notification is missing with the bt_launch changes?
I'll check! Thank you so much!

**Comment by hufman** [comment]
The next time it happens, can you check whether the car's Apps box is still checked? When it happened to me just now, the car showed that the Apps connection had actually disconnected.

**Comment by fabssor** [comment]
Also noticed it does not work smoothly with spotify:
AAiD and MyBMW connects fine when I get in the morning to my car. If I leave work in the afternoon music plays but no coverart is shown. Also I cannot control the music over IDrive. If I force close MyBMW, spotify and AAiD it works agian. I do not have to reconnect to the car. Just have to reopen the apps and everything works just fine.
Next time I am driving I will investigate further and give some more infos!

Maybe this is connected to #217 ?

**Comment by Kamil2398** [comment]
Hi at the beginning, sorry for my English but I am using Google translate: Hufman has the same problem as dominiks66 after some time of listening to music on Spotify, the music suddenly stops playing. The notification remains that the mybmw application is connected, but idrive (idrive6) does not display the connection mark (as if it disconnected them and the question here is whether the matter is on the my bmw application side or whether anroididrive crashing mybmw applications.) In order for the notification from my bmw to disappear, you need to stop the application and restart. I also have a condition described by fabssor. I believe that the problem should be reopened so that it is more visible and more people can see it and contribute to the discussion. If so, I can help, I hope that the language barrier will not be a big problem and everyone will understand what I wanted to convey :)

**Comment by hufman** [comment]
Thank you for all your test results!
However, because the Apps checkbox becomes unchecked, that is a problem at the MyBMW connection level, and AAI can't provide much help there.
The BMWConnected and MyBMW notifications are based on the USB or Bluetooth physical connection, and not directly on the Apps connection itself, which leads to the confusion that the notification is showing without an active connection.
For the purpose of this issue, AAI now supports the North American version of MyBMW and starts up automatically when the phone connects to the car. The issue #244 would be the better place to track any MyBMW connection instability.

**Comment by Kamil2398** [comment]
> Dziękuję za wszystkie wyniki testów!
> Jednak, ponieważ pole wyboru Aplikacje zostaje odznaczone, jest to problem na poziomie połączenia MyBMW, a AAI nie może tam zapewnić wiele pomocy.
> Powiadomienia BMWConnected i MyBMW są oparte na fizycznym połączeniu USB lub Bluetooth, a nie bezpośrednio na samym połączeniu z aplikacjami, co prowadzi do nieporozumień, że powiadomienie jest wyświetlane bez aktywnego połączenia.
> Na potrzeby tego wydania AAI obsługuje teraz północnoamerykańską wersję MyBMW i uruchamia się automatycznie, gdy telefon łączy się z samochodem. Kwestia [nr 244](https://github.com/hufman/AndroidAutoIdrive/issues/244) byłby lepszym miejscem do śledzenia żadnej niestabilności połączenia MyBMW.

And there is no possibility that after some time the AAI application causes an error that crashes Mybmw applications, so that the connection to the tunnel is broken?

**Comment by hufman** [comment]
Indeed, there are actually very few touchpoints from AAI to BMWConnected/MyBMW, which means it has little chance of causing any negative effects:

- BMW Connected starts up any installed CarAPI apps, including AAI, but then immediately detaches from AAI because it doesn't actually implement the CarAPI protocol. This was the primary startup signal before this Bluetooth connection method was added.
- BMWConnected(Classic)/MyBMW runs the BCL tunnel to the car, which simply multiplexes TCP data into the BCL stream to the car. It doesn't parse any of the higher-level Etch protocol or anything, just copies bytes around.
- AAI reaches into the BMWConnected(Classic)/MyBMW APK to load the security module, which is loaded within the AAI process and has no effect on the BMW process.

The only other communications happen in a single direction from BMW to any interested app on the phone, with no response or influence possible:

- BMWConnected(Classic)/MyBMW sends out an ACCESSORY_ATTACHED announcement to the entire phone, which AAI might hear and then start up if its Notification Access is running.
- BMWConnected(Classic)/MyBMW sends out ACCESSORY_INFO reports with debug information about the state of the tunnel, which AAI uses to learn the audio context ID if it missed the original ACCESSORY_ATTACHED announcement. These values are displayed in the advanced connection status.

I tried this test tonight (and you are welcome to confirm too) where I uninstalled AAI and watched the Apps checkbox in the Mobile Devices screen. Instead of driving, I sat watching YouTube while parked 😁 MyBMW was the only app running, no Connected or AAI was installed, and the checkbox was displayed from 19:49 until 20:24, for 35 minutes in this instance. The only influence I could imagine is that AAI would add extra memory pressure and cause Android to kill BMW sooner.

Unfortunately, MyBMW only needs the Apps connection for a short time, and fixing this connection is probably low priority for them. On ID5/6 cars, it runs a small app (not found in BMWConnected) to fetch CDS values to update the car's status in the app, probably for cars without a cloud connection. On ID7 cars, it runs a different protocol to load the Android Auto wifi credentials. I hope BMW does improve this!

**Comment by Kamil2398** [comment]
> Indeed, there are actually very few touchpoints from AAI to BMWConnected/MyBMW, which means it has little chance of causing any negative effects:
> 
> * BMW Connected starts up any installed CarAPI apps, including AAI, but then immediately detaches from AAI because it doesn't actually implement the CarAPI protocol. This was the primary startup signal before this Bluetooth connection method was added.
> * BMWConnected(Classic)/MyBMW runs the BCL tunnel to the car, which simply multiplexes TCP data into the BCL stream to the car. It doesn't parse any of the higher-level Etch protocol or anything, just copies bytes around.
> * AAI reaches into the BMWConnected(Classic)/MyBMW APK to load the security module, which is loaded within the AAI process and has no effect on the BMW process.
> 
> The only other communications happen in a single direction from BMW to any interested app on the phone, with no response or influence possible:
> 
> * BMWConnected(Classic)/MyBMW sends out an ACCESSORY_ATTACHED announcement to the entire phone, which AAI might hear and then start up if its Notification Access is running.
> * BMWConnected(Classic)/MyBMW sends out ACCESSORY_INFO reports with debug information about the state of the tunnel, which AAI uses to learn the audio context ID if it missed the original ACCESSORY_ATTACHED announcement. These values are displayed in the advanced connection status.
> 
> I tried this test tonight (and you are welcome to confirm too) where I uninstalled AAI and watched the Apps checkbox in the Mobile Devices screen. Instead of driving, I sat watching YouTube while parked 😁 MyBMW was the only app running, no Connected or AAI was installed, and the checkbox was displayed from 19:49 until 20:24, for 35 minutes in this instance. The only influence I could imagine is that AAI would add extra memory pressure and cause Android to kill BMW sooner.
> 
> Unfortunately, MyBMW only needs the Apps connection for a short time, and fixing this connection is probably low priority for them. On ID5/6 cars, it runs a small app (not found in BMWConnected) to fetch CDS values to update the car's status in the app, probably for cars without a cloud connection. On ID7 cars, it runs a different protocol to load the Android Auto wifi credentials. I hope BMW does improve this!

I understand that if BMW does not improve the application, it will not be possible to use AAI on long routes and it will be back to Bmw Connected? I have a question dominiks66 reported that version 1.2.2.404-bdec840 is working properly. How it's possible?

**Comment by DominikS66** [comment]
@hufman you just confirmed what I've been afraid of. Problem of broken connection seems to be in actual MyBMW app. @Kamil2398 I'm also wondering...latest test yesterday had now also the disconnection after arround 35 minutes. The timespan seems to be a little bit random. Maybe we have two problems here: On my first report of Sunday, I need to restart Spotify because AAID lost control to Spotify. Only after restarting both Apps the connection from AAId to Car came back. On latest test with disconnection after 35 minutes, just switching BT off and on in iDrive 6 brought back the connection to AAId (I assume this action restarts MyBMW car connection). No further action was required on my phone.

**Comment by Kamil2398** [comment]
I post pictures showing the problem with the Spotify API, the question is whether I should start a new thread related to this?
![20210414_161316](https://user-images.githubusercontent.com/82474154/114728315-47eac200-9d3f-11eb-9593-31253a2df0e8.jpg)
![Screenshot_20210414-161359_AndroidAutoIdrive](https://user-images.githubusercontent.com/82474154/114728396-576a0b00-9d3f-11eb-9b9f-31af08bcfd8d.jpg)
This occurred after I left work and properly connected, the music was playing in the background

**Comment by hufman** [comment]
The times I experienced the disconnect, I was able to disconnect and connect the Bluetooth Music connection from within IDrive to bring back the Apps connection.

The screenshots about Spotify are interesting. The phone UI player doesn't interact with BMW at all, so it would better fit in a separate issue. You should only need to kill Spotify, and maybe switch apps within AAI to reconnect. Granting Notification Access should add an additional method to control the music app even if the Spotify API is down, though I think Spotify doesn't reliably show coverart this way. This would use the same protocol as your phone's lock screen to show and control the music, does Spotify's information properly show up on the lock screen in this state?

**Comment by unlive74** [comment]
I got the same problem as @Kamil2398. And it seems to me that there was no such problem before:

> The phone UI player doesn't interact with BMW at all, so it would better fit in a separate issue. You should only need to kill Spotify, and maybe switch apps within AAI to reconnect.

Perhaps something has changed in the spotify interface..

**Comment by LACNR** [comment]
Maybe try out the old Connected 6.4, which was meant to support apps. MyBMW officially doesn't support apps in the car and the app recognition i iDrive7 works different, remember on Apple you don't even need a BMW App to use apps in-car.

Could be Spotify itself but I also believe because MyBMW isn't meant to add apps so maybe the tunnel that is open isn't properly/fully in service.

**Comment by unlive74** [comment]
@LACNR I use BMW Connected 6.4

**Comment by LACNR** [comment]
Okay.

I have iDrive4++, Connected 6.4 (and MyBMW) installed and everything works fine. I use Spotify everday, though it might be important I haven't updated the Spotify App since september last year.

**Comment by unlive74** [comment]
By the way, yes. I will try to roll back to the old version of spotify, there may be a problem with this

**Comment by Kamil2398** [comment]
AAI has all the permissions granted to me when everything works, it is checked 3x. When this failure occurred, the check was x2, the middle one was missing, then the check and the authorization option was available. I hope you will understand what I mean 😁. The problem probably occurred after one of the new Spotify updates, I have not experienced it before. Hufman, is there a chance to write to bmw to fix Mybmw so that it does not disconnect after a certain longer period and you do not have to restart the connection?

**Comment by LACNR** [comment]
You could write down a connectivity issue to ConnectedDrive deaprtment or BMW main via the MyBMW app email listed in the play store or comment in the play store. I did a week ago because MyBMW doesn't show me the trip information in the app, but haven't heard of them yet. They usually call me personally on my phone.

**Comment by LACNR** [comment]
> You could write down a connectivity issue to ConnectedDrive deaprtment or BMW main via the MyBMW app email listed in the play store or comment in the play store. Though MyBMW isn't yet planned to integrate apps so I'm not sure if they focuse on a consistent connection if it is only meant to update trip data.

I did a week ago because MyBMW doesn't show me the trip information in the app, but haven't heard of them yet. They usually call me personally on my phone.

**Comment by Kamil2398** [comment]
I sent the report, the more people do that, the more likely it is to improve. I installed Spotify on April 1 version and will be testing it

**Comment by dimakaob** [comment]
If the problem is Spotify here, so we can try Youtube music and see how it works?

**Comment by LACNR** [comment]
You technically could try out any other app I guess even the standard music player on your phone (e.g. Samsung Music Player)

**Comment by LACNR** [comment]
There's a new update for MyBMW on the PlayStore (from April 15th although available only since now)

**Comment by Kamil2398** [comment]
I think we should move all discussions to #244 Interestingly, for me, version 1.34 is not available for update yet.
